internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,304 @@
+	if err != nil {
+		return nil, fmt.Errorf("delegating_resolver: failed to determine proxy URL for target %s: %v", target, err)
+	}
+
+	// proxy is not configured or proxy address excluded using `NO_PROXY` env
+	// var, so only target resolver is used.
+	if r.proxyURL == nil {
+		return targetResolverBuilder.Build(target, cc, opts)
+	}
+
+	if logger.V(2) {
+		logger.Info("Proxy URL detected : %s", r.proxyURL)
+	}
+
+	// When the scheme is 'dns' and target resolution on client is not enabled,
+	// resolution should be handled by the proxy, not the client. Therefore, we
+	// bypass the target resolver and store the unresolved target address.
+	if target.URL.Scheme == "dns" && !targetResolutionEnabled {
+		r.targetAddrs = []resolver.Address{{Addr: target.Endpoint()}}
+		r.targetResolverReady = true
+	} else {
+		wcc := &wrappingClientConn{
+			parent:       r,
+			resolverType: targetResolverType,
+		}
+		if r.targetResolver, err = targetResolverBuilder.Build(target, wcc, opts); err != nil {
+			return nil, fmt.Errorf("delegating_resolver: unable to build the resolver for target %s: %v", target, err)
+		}
+	}
+
+	if r.proxyResolver, err = r.proxyURIResolver(opts); err != nil {
+		return nil, fmt.Errorf("delegating_resolver: failed to build resolver for proxy URL %q: %v", r.proxyURL, err)
+	}
+	return r, nil
+}
+
+// proxyURIResolver creates a resolver for resolving proxy URIs using the
+// "dns" scheme. It adjusts the proxyURL to conform to the "dns:///" format and
+// builds a resolver with a wrappingClientConn to capture resolved addresses.
+func (r *delegatingResolver) proxyURIResolver(opts resolver.BuildOptions) (resolver.Resolver, error) {
+	proxyBuilder := resolver.Get("dns")
+	if proxyBuilder == nil {
+		panic(fmt.Sprintln("delegating_resolver: resolver for proxy not found for scheme dns"))
+	}
+	r.proxyURL.Scheme = "dns"
+	r.proxyURL.Path = "/" + r.proxyURL.Host
+	r.proxyURL.Host = "" // Clear the Host field to conform to the "dns:///" format
+
+	proxyTarget := resolver.Target{URL: *r.proxyURL}
+	wcc := &wrappingClientConn{
+		parent:       r,
+		resolverType: proxyResolverType,
+	}
+	return proxyBuilder.Build(proxyTarget, wcc, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+	}
+}
+
+// generateCombinedAddressesLocked creates a list of combined addresses by
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) generateCombinedAddressesLocked() ([]resolver.Address, []resolver.Endpoint) {
```

Comment:
nit: Can omit `generate` from the name. Go functions are names start with nouns instead of verbs, unless there is some asynchronous work being done.
https://google.github.io/styleguide/go/decisions#getters